shark_turbine/kernel/wave/minimize_global_loads.py

```diff
@@ -0,0 +1,237 @@
+    custom = get_custom(node)
+    return (
+        isinstance(custom, Read)
+        and custom.type.address_space.subs(idxc.subs) == GLOBAL_ADDRESS_SPACE
+        and has_write_shared_user(custom)
+    )
+
+
+def construct_min_global_access_pattern(
+    index: dict[IndexSymbol, IndexSequence],
+    thread_id: IndexExpr,
+    load_elems_per_thread: int,
+    shape: list[int],
+) -> dict[IndexSymbol, IndexSequence]:
+    """
+    This function constructs a new access pattern for a global read node.
+    It retains workgroup and induction variable indexing but removes any thread
+    level indexing which is inherited from the mma nodes during expansion.
+    It takes a 1-D global offset and delinearizes it to a multi-dimensional offset
+    and updates the access pattern accordingly.
+    """
+    thread_ids = [THREAD_0, THREAD_1, THREAD_2]
+    new_index = {key: index[key].subs({t: 0 for t in thread_ids}) for key in index}
+    nd_index = delinearize_index(thread_id, shape)
+    for i, key in enumerate(index.keys()):
+        new_index[key].start += nd_index[i]
+        new_index[key].size = load_elems_per_thread
+    return new_index
+
+
+def remove_global_indexing(
+    index: dict[IndexSymbol, IndexSequence], tilingConstraints: list[TilingConstraint]
+) -> dict[IndexSymbol, IndexSequence]:
+    """
+    This function takes the index sequence for a global read and removes all
+    workgroup and induction level indexing. This is necessary for writes to shared memory
+    that operate on promoted memory.
+    """
+    workgroup_ids = [WORKGROUP_0, WORKGROUP_1, WORKGROUP_2]
+    new_index = {key: index[key].subs({w: 0 for w in workgroup_ids}) for key in index}
+    for key in new_index:
+        for constraint in tilingConstraints:
+            new_index[key] = new_index[key].subs({constraint.induction_var: 0})
+    return new_index
+
+
+def materialize_shape(
+    constraint_tile_size: dict[IndexSymbol, int], symbolic_shape: list[IndexSymbol]
+) -> list[int]:
+    materialized_shape = []
+    idxc = IndexingContext.current()
+    for dim in symbolic_shape:
+        if dim in constraint_tile_size:
+            materialized_shape.append(constraint_tile_size[dim].subs(idxc.subs))
+        else:
+            materialized_shape.append(dim.subs(idxc.subs))
+    return materialized_shape
+
+
+def identify_optimizable_loads(
+    global_read_nodes: list[fx.Node],
+    constraint_tile_size: dict[IndexSymbol, int],
+    max_elements_per_load: int,
+) -> list[Read]:
+    """
+    Identify sub-optimal global loads that can be removed. A given memory has
+    sub-optimal global loads if
+        num_global_loads > (M * N) / (T * L)
+    where the memory has shape [M, N], there are T threads and each thread can load L elements.
+    """
+    optimizable_loads: dict[fx.Node, tuple[int, Read]] = {}
+    for read_node in global_read_nodes:
+        custom = get_custom(read_node)
+        if custom.memory in optimizable_loads:
```

Comment:
So we are collecting/indexing on the global memory we are loading from as opposed to tevery instance of readOps right?